File: best_time_to_buy_and_sell_stock.cpp
/*
121. Best Time to Buy and Sell Stock

Problem Statement:
You are given an array 'prices' where prices[i] represents the price of a given stock on the i-th day.
You need to determine the maximum profit that can be achieved by buying on one day and selling on a later day.
If no profit can be made, return 0.

Example 1:
Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.

Example 2:
Input: prices = [7,6,4,3,1]
Output: 0
Explanation: No transactions are done, max profit = 0.

Constraints:
1 <= prices.length <= 10^5
0 <= prices[i] <= 10^4

Approach:
1. Initialize `min` to the first element of `prices`, representing the minimum price encountered so far.
2. Initialize `maxProf` to 0, representing the maximum profit found so far.
3. Traverse the array:
   - Calculate the profit if we sell on the current day: `profit = prices[i] - min`.
   - If `prices[i]` is lower than `min`, update `min` (as we want to buy at the lowest price).
   - If `profit` is greater than `maxProf`, update `maxProf`.
4. Return `maxProf` as the result.

Time Complexity: O(n) (Single pass through the array)
Space Complexity: O(1) (Constant extra space used)
*/

class Solution {
  public:
      int maxProfit(vector<int>& prices) {
          int n = size(prices);
          int min = prices[0];  // Stores the minimum price encountered so far
          int maxProf = 0;  // Stores the maximum profit
  
          for(int i = 0; i < n; i++) {
              int prof = prices[i] - min;  // Calculate profit if selling on day i
              
              if(prices[i] < min) {  
                  min = prices[i];  // Update min price if a lower price is found
              }
              
              if(prof > maxProf) {  
                  maxProf = prof;  // Update max profit if a higher profit is found
              }
          }
         
          return maxProf;  // Return the maximum profit found
      }
  };
  

File: Array/best_time_to_buy_and_sell_stock.js
/**
 * 📈 Problem: Best Time to Buy and Sell Stock
 *
 * Given an array prices where prices[i] is the price of a stock on day i,
 * find the maximum profit you can achieve.
 *
 * You may buy on one day and sell on a different future day.
 * If no profit is possible, return 0.
 */

/**
 * 🔴 BRUTE FORCE APPROACH
 *
 * Idea:
 * - Try every possible buy day (i)
 * - Try every possible sell day after it (j)
 * - Calculate profit = prices[j] - prices[i]
 * - Track the maximum profit
 *
 * Time Complexity (TC): O(n²)
 * Space Complexity (SC): O(1)
 *
 * ❌ Inefficient for large inputs
 */
var maxProfitBruteForce = function (prices) {
    let maxProfit = 0;

    for (let i = 0; i < prices.length; i++) {
        for (let j = i + 1; j < prices.length; j++) {
            let profit = prices[j] - prices[i];
            maxProfit = Math.max(maxProfit, profit);
        }
    }
    return maxProfit;
};

/**
 * 🟢 OPTIMAL APPROACH (Greedy)
 *
 * 🔑 Pattern Used:
 * - Greedy Algorithm
 * - Minimum Tracking (Prefix Minimum)
 * - Buy Low, Sell High (Single Transaction)
 *
 * Idea:
 * - Track the minimum price so far (best day to buy)
 * - At each day, calculate profit if sold today
 * - Update maximum profit
 *
 * Formula:
 * profit = currentPrice - minPriceSoFar
 *
 * Time Complexity (TC): O(n)
 * Space Complexity (SC): O(1)
 *
 * ✅ Best solution (used in interviews)
 */
var maxProfitOptimal = function (prices) {
    let minPrice = Infinity; // best price to buy so far
    let maxProfit = 0; // best profit so far

    for (let price of prices) {
        // Update minimum price
        minPrice = Math.min(minPrice, price);

        // Calculate profit if sold today
        let profit = price - minPrice;

        // Update maximum profit
        maxProfit = Math.max(maxProfit, profit);
    }
    return maxProfit;
};

/**
 * 🧠 PATTERN SUMMARY (For Revision)
 *
 * Pattern Name:
 * - Greedy + Minimum Tracking
 *
 * When to Use:
 * - Buy before sell
 * - Maximum difference problem
 * - One transaction allowed
 *
 * Key Insight:
 * - Always buy at the lowest price seen so far
 * - Always try to sell at the current price
 *
 * Mental Shortcut:
 * max(A[j] - A[i]) where j > i
 * → Think Greedy + Prefix Minimum
 */
